LibertyDSNP/frequency
# Feat/separate unlock chunks #1753

# Goal
The goal of this PR is to separate unstaking unlock chunks into their own storage and do a storage migration which includes staking type, for convenience of the planned Provider Boost feature so that another storage migration is not needed for that feature.  There are multiple advantages to this separation discussed in the linked issue.

Closes #1753 .  Please see the linked issue for more details about this proposed change.

# Discussion

A stake will require 1 extra read of unstaking unlocks from before, in order to set the account token lock correctly.  This was not noted in the original issue.

As noted in the original issue, an unstake will require 1 extra read and 1 extra write from before.

Effectively changes the name of `StakingAccountDetails` to just `StakingDetails` while splitting the storage.

`StakingAccountLedger` now stores `StakingDetails`.

The migration uses the `translate` function of the `IterableStorageMap` trait.

The implementation design document was updated to make it more clear that details of actual ProviderBoost staking rewards are undecided and unspecified.


Since token locks are being touched, please review with an eye toward bugs with token locks and test coverage.

## To Verify
[Install the CLI version of try-runtime](https://paritytech.github.io/try-runtime-cli/try_runtime/#installation), then run try-runtime to test the migration against Frequency Rococo:
```shell
cargo build --release --features frequency-rococo-testnet,try-runtime && \
try-runtime --runtime ./target/release/wbuild/frequency-runtime/frequency_runtime.wasm on-runtime-upgrade live --uri wss://rpc.rococo.frequency.xyz:443 -p Capacity
```

You should see at something like:
```
[2023-11-15T01:32:16Z INFO  runtime::capacity] 🔄 StakingAccountLedger migration started
[2023-11-15T01:32:16Z INFO  runtime::capacity] migrated 1
[2023-11-15T01:32:16Z INFO  runtime::capacity] migrated 2
...
[2023-11-15T01:32:16Z INFO  runtime::capacity] migrated 335
[2023-11-15T01:32:16Z INFO  runtime::capacity] migrated 336
[2023-11-15T01:32:16Z INFO  runtime::capacity] 🔄 migration finished
[2023-11-15T01:32:16Z INFO  runtime::capacity] Migration calculated weight = Weight { ref_time: 75725000000, proof_size: 0 }
[2023-11-15T01:32:16Z INFO  runtime::capacity] ✅ migration post_upgrade checks passed
```
The errors afterward are unrelated to the migration, having to do with a runtime difference between local and rococo, and can be ignored.

# Checklist
- [x] Storage migration added
- [x] Storage migration tested via try_runtime
- [x] <del>Custom RPC OR Runtime API added/changed? Updated js/api-augment.</del>
- [x] Benchmarks updated (need to add requested changes)
- [x] Weights updated (need to rerun with requested changes)
